server/s3_multiobject_copy_action.cc

```diff
@@ -0,0 +1,1275 @@
+      request, object_multipart_metadata->get_part_index_layout(), upload_id,
+      part_number);
+
+  part_metadata->load(
+      std::bind(&S3MultiObjectCopyAction::fetch_part_info_success, this),
+      std::bind(&S3MultiObjectCopyAction::fetch_part_info_failed, this),
+      part_number);
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3MultiObjectCopyAction::fetch_part_info_success() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (part_metadata->get_state() == S3PartMetadataState::present) {
+    s3_log(S3_LOG_DEBUG, request_id,
+           "S3PartMetadataState::present for part %d\n", part_number);
+    old_object_oid = part_metadata->get_oid();
+    old_oid_str = S3M0Uint128Helper::to_string(old_object_oid);
+    old_layout_id = part_metadata->get_layout_id();
+    next();
+  } else if (part_metadata->get_state() == S3PartMetadataState::missing) {
+    s3_log(S3_LOG_DEBUG, request_id, "S3PartMetadataState::missing\n");
+    next();
+  } else if (part_metadata->get_state() ==
+             S3PartMetadataState::failed_to_launch) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Part metadata load operation failed due to pre launch failure\n");
+    s3_copy_part_action_state = S3CopyPartActionState::validationFailed;
+    set_s3_error("ServiceUnavailable");
+    send_response_to_s3_client();
+  } else {
+    s3_log(S3_LOG_DEBUG, request_id, "Failed to look up metadata.\n");
+    s3_copy_part_action_state = S3CopyPartActionState::validationFailed;
+    set_s3_error("InternalError");
+    send_response_to_s3_client();
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3MultiObjectCopyAction::fetch_part_info_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (part_metadata->get_state() == S3PartMetadataState::missing) {
+    next();
+  } else {
+    s3_copy_part_action_state = S3CopyPartActionState::validationFailed;
+    set_s3_error("InternalError");
+    send_response_to_s3_client();
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+bool S3MultiObjectCopyAction::copy_object_cb() {
+  if (check_shutdown_and_rollback() || !request->client_connected()) {
+    return true;
+  }
+  if (response_started) {
+    request->send_reply_body(xml_spaces, sizeof(xml_spaces) - 1);
+  }
+  return false;
+}
+
+void S3MultiObjectCopyAction::create_target_object() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (additional_object_metadata->get_number_of_fragments() == 0) {
+    create_object();
+  } else {
+    // Incase of fragments or multipart upload
+    // each part copy needs separate object creation
+    create_objects();
+  }
+  s3_log(S3_LOG_DEBUG, stripped_request_id, "%s Exit", __func__);
+}
+
+void S3MultiObjectCopyAction::create_object() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  s3_timer.start();
+  motr_writer = motr_writer_factory->create_motr_writer(
```

Comment:
this code may not be exact one which we may need to use !? As seen in CopyObject API? and in 
https://github.com/Seagate/cortx-s3server/pull/1636/files#diff-f35bcc9236fe541dc5b3914cce0dea32790f77664d9eb362938182c0e933a402R293

check for similar code elsewhere if its suitable

Reply:
Function definition is same, but name should be different as we are dealing with parts and not full object. For simple copy it is create_object()/create_objects() and for part-copy it should be create_part()/create_parts().